sbnci/Modules/TrackValidation/TrackValidation_module.cc

```diff
@@ -0,0 +1,3127 @@
+  }
+
+  PFProjectionMatched_TreeVal[fTrackModuleLabel].push_back(projectionMatched);
+  if(pfp->PdgCode() == 11){PFPShowerProjectionMatched_TreeVal[fTrackModuleLabel].push_back(projectionMatched);
+  }else if(pfp->PdgCode() == 13){PFPTrackProjectionMatched_TreeVal[fTrackModuleLabel].push_back(projectionMatched);}
+
+
+  if(totalhits != 0){
+    completeness_hits = (signalhits)/totalhits;
+  }//close if statement
+
+  if(pfphits != 0){
+    purity_hits = signalhits/pfphits;
+  }
+
+  if(TotalTrueEnergy != 0){
+    completeness_energy = pfpenergy/TotalTrueEnergy;
+  }
+
+  if(TotalEnergyDepinHits != 0){
+    purity_energy = pfpenergy/TotalEnergyDepinHits;
+  }
+
+  PFPHitsCompl_TreeVal[fTrackModuleLabel].push_back(completeness_hits);
+  PFPHitsPurity_TreeVal[fTrackModuleLabel].push_back(purity_hits);
+  PFPEnergyCompl_TreeVal[fTrackModuleLabel].push_back(completeness_energy);
+  PFPEnergyPurity_TreeVal[fTrackModuleLabel].push_back(purity_energy);
+  //For Tracks
+  if(pfp->PdgCode() == 13){
+    PFPTrackHitsCompl_TreeVal[fTrackModuleLabel].push_back(completeness_hits);
+    PFPTrackHitsPurity_TreeVal[fTrackModuleLabel].push_back(purity_hits);
+    PFPTrackEnergyCompl_TreeVal[fTrackModuleLabel].push_back(completeness_energy);
+    PFPTrackEnergyPurity_TreeVal[fTrackModuleLabel].push_back(purity_energy);
+  }else if(pfp->PdgCode() == 11){
+    PFPShowerHitsCompl_TreeVal[fTrackModuleLabel].push_back(completeness_hits);
+    PFPShowerHitsPurity_TreeVal[fTrackModuleLabel].push_back(purity_hits);
+    PFPShowerEnergyCompl_TreeVal[fTrackModuleLabel].push_back(completeness_energy);
+    PFPShowerEnergyPurity_TreeVal[fTrackModuleLabel].push_back(purity_energy);
+  }
+
+  if(fVerbose>1){
+    std::cout << "#################################################" << std::endl;
+    std::cout << "                 PFP Metrics                     " << std::endl;
+    std::cout << "#################################################" << std::endl;
+    std::cout << "Projection matched:      " << projectionMatched    << std::endl;
+    std::cout << "PFP hit completeness:    " << completeness_hits    << std::endl;
+    std::cout << "PFP hit purity:          " << purity_hits          << std::endl;
+    std::cout << "PFP energy completeness: " << completeness_energy  << std::endl;
+    std::cout << "PFP energy purity:       " << purity_energy        << std::endl;
+    std::cout << "#################################################" << std::endl;
+
+  }//close fverbose
+  return;
+}//close phparticel validation
+
+
+//
+//
+//#################################################################################################
+//
+//=========================================================
+//======= These function can be moved to TrackUtils =======
+//=========================================================
+
+//Function retunr a map of ShowerMother trackid as a key and a vector of daughters ids
+void ana::TrackValidation::ShowerMotherTrackIdsFun(art::Event const& evt,
```

Comment:
Isn't this code a duplicate of ShowerUtils?https://github.com/SBNSoftware/sbnci/blob/0a0b710a85d9970e37d8c15cf8539dd2cbee88b5/sbnci/Modules/MCRecoUtils/ShowerUtils.cc#L3

Reply:
resolved